ibis/expr/types/arrays.py

```diff
@@ -1108,15 +1116,49 @@ def array(values: Iterable[V]) -> ArrayValue:
     │ [3, 42, ... +1]      │
     └──────────────────────┘
 
-    >>> ibis.array([t.a, 42 + ibis.literal(5)])
+    >>> ibis.array([t.a, 42 + ibis.literal(5)], type="array<float>")
     ┏━━━━━━━━━━━━━━━━━━━━━━┓
     ┃ Array()              ┃
     ┡━━━━━━━━━━━━━━━━━━━━━━┩
-    │ array<int64>         │
+    │ array<float64>       │
     ├──────────────────────┤
-    │ [1, 47]              │
-    │ [2, 47]              │
-    │ [3, 47]              │
+    │ [1.0, 47.0]          │
+    │ [2.0, 47.0]          │
+    │ [3.0, 47.0]          │
     └──────────────────────┘
     """
-    return ops.Array(tuple(values)).to_expr()
+    type = dt.dtype(type) if type is not None else None
+    if type is not None and not isinstance(type, dt.Array):
+        raise ValidationError(f"type must be an array, got {type}")
+
+    if isinstance(values, ir.Value):
+        if type is not None:
+            return values.cast(type)
+        elif isinstance(values, ArrayValue):
+            return values
+        else:
+            raise ValidationError(
+                f"If no type passed, values must be an array, got {values.type()}"
+            )
+
+    if values is None:
+        if type is None:
+            raise ValidationError("If values is None/NULL, type must be provided")
+        return ir.null(type)
+
+    values = tuple(values)
+    if len(values) == 0:
+        if type is None:
+            raise ValidationError("If values is empty, type must be provided")
+        return ops.EmptyArray(type).to_expr()
+    else:
+        value_type = type.value_type if type is not None else None
+        values = [_value(v, value_type) for v in values]
+        return ops.Array(values).to_expr()
+
+
+def _value(x, type) -> ir.Value:
+    if isinstance(x, (ir.Value, Deferred)):
+        return x.cast(type) if type is not None else x
+    else:
+        return ibis.literal(x, type=type)
```

Comment:
What is the purpose of `literal` here?

Reply:
If you do ibis.array([1, 2], type="array<float>"), we somehow need to get those ints to floats. My original implementation stored them in the array as ints, and convert to float during compilation. But that required storing the eventual type in ops.Array. so here we pre convert all the elements to floats, and now ops.Array can just use rlz.highest_precedence_dtype(). In order to do that pre casting, we need this literal.

Reply:
I'm still not sure why this is required. Is there some API that fails to execute if you don't do these casts?

What if the elements themselves are arrays? Doesn't this then need to be recursive in the casting/call to `_value`?

Reply:
Shoot, github escaped the `<>` I posted, I meant `ibis.array([1, 2], type="array<float>")` not `ibis.array([1, 2], type="array")`

Now does the concern make sense? Two ways of storing this:
1. store the inputs as is, and keep track of the dtype separately `op.Array(values=[1,2], value_type="float")`
2. cast the inputs early, then we don't need to store the dtype: `op.Array(values=[op.Literal(1, "float"), op.Literal(2, "float")])`

I am trying to go for option 2.

> I'm still not sure why this is required

Specificaly, if I remove the .literal() here, then we pass a plain python `1` into the op.Array constructor, and then using pattern matching it tries to coerce that to an ibis expression, which yields a `op.Literal(1, int)`. But we wanted it to be a float, so we have to do the casting up front here.

> What if the elements themselves are arrays

you mean what if x were a python list of ibis values? ie someone did `array_of_arrays = ibis.array([[ibis.literal(1)], [ibis.literal(2)]])`? Yes, this would fail, good catch. not sure what to do about it yet....